pkg/runtime/types.go

```diff
@@ -120,26 +132,370 @@ type Unknown struct {
+				strMap[k] = str
+			}
+		}
+		return strMap
+	}
+	return nil
+}
+
+func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+	m := obj
+	if len(fields) > 1 {
+		for _, field := range fields[0 : len(fields)-1] {
+			if _, ok := m[field].(map[string]interface{}); !ok {
+				m[field] = make(map[string]interface{})
+			}
+			m = m[field].(map[string]interface{})
+		}
+	}
+	m[fields[len(fields)-1]] = value
+}
+
+func setNestedMap(obj map[string]interface{}, value map[string]string, fields ...string) {
+	m := make(map[string]interface{}, len(value))
+	for k, v := range value {
+		m[k] = v
+	}
+	setNestedField(obj, m, fields...)
+}
+
+func (u *Unstructured) setNestedField(value interface{}, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedField(u.Object, value, fields...)
+}
+
+func (u *Unstructured) setNestedMap(value map[string]string, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedMap(u.Object, value, fields...)
+}
+
+// TODO: rewrite this.
+func setField(src reflect.Value, fieldName string, dest interface{}) error {
+	field := src.FieldByName(fieldName)
+	if !field.IsValid() {
+		return fmt.Errorf("%v does not have field %s", src, fieldName)
+	}
+	v, err := conversion.EnforcePtr(dest)
+	if err != nil {
+		return err
+	}
+	if field.Type().AssignableTo(v.Type()) {
+		v.Set(field)
+		return nil
+	}
+	if field.Type().ConvertibleTo(v.Type()) {
+		v.Set(field.Convert(v.Type()))
+		return nil
+	}
+	return fmt.Errorf("couldn't assign/convert %v to %v", field.Type(), v.Type())
+}
+
+func extractOwnerReference(src interface{}) (metatypes.OwnerReference, error) {
+	var ret metatypes.OwnerReference
+	v := src.(map[string]interface{})
+	ret.Kind = getNestedString(v, "kind")
+	ret.Name = getNestedString(v, "name")
+	ret.APIVersion = getNestedString(v, "apiVersion")
+	ret.UID = (types.UID)(getNestedString(v, "uid"))
+	return ret, nil
+}
+
+func setOwnerReference(dest interface{}, src metatypes.OwnerReference) error {
+	v, ok := dest.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("dest %v has to be of type map[string]interface()", dest)
+	}
+	setNestedField(v, src.Kind, "kind")
+	setNestedField(v, src.Name, "name")
+	setNestedField(v, src.APIVersion, "apiVersion")
+	setNestedField(v, string(src.UID), "uid")
+	return nil
+}
+
+func getOwnerReferences(object map[string]interface{}) (ownerReferences []map[string]interface{}, err error) {
+	field := getNestedField(object, "metadata", "ownerReferences")
+	if field == nil {
+		return ownerReferences, fmt.Errorf("cannot find field metadata.ownerReferences in %v", object)
+	}
+	ownerReferences, ok := field.([]map[string]interface{})
+	if !ok {
+		return ownerReferences, fmt.Errorf("expect metadata.ownerReferences to be a slice in %v", object)
+	}
+	return
+}
+
+func (u *Unstructured) GetOwnerReference(i int) (metatypes.OwnerReference, error) {
```

Comment:
Just for reference if you remove the pointers from the arrays in @krousey's
benchmark code those alloc numbers should look much more reasonable. You
don't need an array of pointers anyway.

On Thu, May 5, 2016 at 10:29 AM, Chao Xu <notifications@github.com> wrote:

> In pkg/runtime/types.go
> <https://github.com/kubernetes/kubernetes/pull/24509#discussion_r62221980>
> :
>
> > +	return nil
> > +}
> > +
> > +func getOwnerReferences(object map[string]interface{}) (ownerReferences []map[string]interface{}, err error) {
> > +	field := getNestedField(object, "metadata", "ownerReferences")
> > +	if field == nil {
> > +		return ownerReferences, fmt.Errorf("cannot find field metadata.ownerReferences in %v", object)
> > +	}
> > +	ownerReferences, ok := field.([]map[string]interface{})
> > +	if !ok {
> > +		return ownerReferences, fmt.Errorf("expect metadata.ownerReferences to be a slice in %v", object)
> > +	}
> > +	return
> > +}
> > +
> > +func (u *Unstructured) GetOwnerReference(i int) (metatypes.OwnerReference, error) {
>
> Last two columns are memory numbers.
>
>           AccessAllDirect(  1): (10000000          133 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy(  1): (10000000          235 ns/op)(      72 B/op        2 allocs/op)
>              RemoveDirect(  1): (100000000          28.4 ns/op)(       0 B/op          0 allocs/op)
>            RemoveWithCopy(  1): ( 5000000          359 ns/op)(     104 B/op        3 allocs/op)
>              AppendDirect(  1): (20000000           95.6 ns/op)(      16 B/op          1 allocs/op)
>            AppendWithCopy(  1): ( 1000000         1859 ns/op)(     936 B/op       17 allocs/op)
> --------------
>           AccessAllDirect(  2): ( 5000000          270 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy(  2): ( 3000000          428 ns/op)(     144 B/op        3 allocs/op)
>              RemoveDirect(  2): (10000000          140 ns/op)(       0 B/op        0 allocs/op)
>            RemoveWithCopy(  2): ( 1000000         1264 ns/op)(     584 B/op       11 allocs/op)
>              AppendDirect(  2): (10000000          141 ns/op)(      32 B/op        1 allocs/op)
>            AppendWithCopy(  2): (  500000         2981 ns/op)(    1440 B/op       24 allocs/op)
> --------------
>           AccessAllDirect(  5): ( 2000000          634 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy(  5): ( 2000000          980 ns/op)(     368 B/op        6 allocs/op)
>              RemoveDirect(  5): ( 5000000          269 ns/op)(       0 B/op        0 allocs/op)
>            RemoveWithCopy(  5): (  500000         4018 ns/op)(    2032 B/op       32 allocs/op)
>              AppendDirect(  5): (10000000          160 ns/op)(      80 B/op        1 allocs/op)
>            AppendWithCopy(  5): (  300000         5326 ns/op)(    2928 B/op       45 allocs/op)
> --------------
>           AccessAllDirect( 10): ( 1000000         1221 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy( 10): ( 1000000         1845 ns/op)(     720 B/op       11 allocs/op)
>              RemoveDirect( 10): ( 2000000          626 ns/op)(       0 B/op        0 allocs/op)
>            RemoveWithCopy( 10): (  200000         8189 ns/op)(    4432 B/op       67 allocs/op)
>              AppendDirect( 10): (10000000          183 ns/op)(     160 B/op        1 allocs/op)
>            AppendWithCopy( 10): (  200000         9580 ns/op)(    5408 B/op       80 allocs/op)
> --------------
>           AccessAllDirect( 20): (  500000         2448 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy( 20): (  500000         3429 ns/op)(    1440 B/op       21 allocs/op)
>              RemoveDirect( 20): ( 1000000         1206 ns/op)(       0 B/op        0 allocs/op)
>            RemoveWithCopy( 20): (  100000        16366 ns/op)(    9232 B/op      137 allocs/op)
>              AppendDirect( 20): ( 5000000          229 ns/op)(     320 B/op        1 allocs/op)
>            AppendWithCopy( 20): (  100000        18313 ns/op)(   10368 B/op      150 allocs/op)
> --------------
>           AccessAllDirect( 50): (  200000         6155 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy( 50): (  200000         8948 ns/op)(    3616 B/op       51 allocs/op)
>              RemoveDirect( 50): (  500000         3282 ns/op)(       0 B/op        0 allocs/op)
>            RemoveWithCopy( 50): (   30000        44495 ns/op)(   23664 B/op      347 allocs/op)
>              AppendDirect( 50): ( 3000000          440 ns/op)(     896 B/op        1 allocs/op)
>            AppendWithCopy( 50): (   30000        44313 ns/op)(   25360 B/op      360 allocs/op)
> --------------
>           AccessAllDirect(100): (  100000        12115 ns/op)(       0 B/op        0 allocs/op)
>             AccessAllCopy(100): (  100000        17599 ns/op)(    7296 B/op      101 allocs/op)
>              RemoveDirect(100): (  200000         6182 ns/op)(       0 B/op        0 allocs/op)
>            RemoveWithCopy(100): (   20000        91465 ns/op)(   47825 B/op      697 allocs/op)
>              AppendDirect(100): ( 2000000          629 ns/op)(    1664 B/op        1 allocs/op)
>            AppendWithCopy(100): (   20000        92102 ns/op)(   50289 B/op      710 allocs/op)
> --------------
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24509/files/a589b65c4ba566ee1ee7558ba3a38d96ca6302e7#r62221980>
>